Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1068 @@
+                goto error;
+            }
+            stack_pointer = ctx->frame->stack_pointer;
+            // Cannot determine statically, so we can't propagate types.
+            if (!sym_is_type(self_or_null, SELF_OR_NULL)) {
+                for (int i = 0; i < argcount; i++) {
+                    ctx->frame->locals[i] = args[i];
                 }
-                break;
             }
-            case _PUSH_FRAME:
-            case _POP_FRAME:
-                co = (PyCodeObject *)buffer[pc].operand;
-                break;
-            case _JUMP_TO_TOP:
-            case _EXIT_TRACE:
-                return;
+            break;
+        }
+
+        case _POP_FRAME: {
+            assert(STACK_LEVEL() == 1);
+            _Py_UOpsSymType *retval = PEEK(1);
+            STACK_SHRINK(1);
+            ctx->frame->stack_pointer = stack_pointer;
+
+            if (ctx_frame_pop(ctx) != 0){
+                goto error;
+            }
+            stack_pointer = ctx->frame->stack_pointer;
+            // Push retval into new frame.
+            STACK_GROW(1);
+            PEEK(1) = retval;
+            break;
         }
+
+        case _SET_IP:
+        case _CHECK_VALIDITY:
+        case _SAVE_RETURN_OFFSET:
+            break;
+        default:
+            DPRINTF(1, "Unknown opcode in abstract interpreter\n");
+            Py_UNREACHABLE();
     }
+    assert(ctx->frame != NULL);
+    DPRINTF(3, " stack_level %d\n", STACK_LEVEL());
+    ctx->frame->stack_pointer = stack_pointer;
+    assert(STACK_LEVEL() >= 0);
+
+    if (emit_i(&ctx->emitter, new_inst) < 0) {
+        return ABSTRACT_INTERP_ERROR;
+    }
+
+    return ABSTRACT_INTERP_NORMAL;
+
+pop_2_error_tier_two:
+    STACK_SHRINK(1);
+    STACK_SHRINK(1);
+error:
+    DPRINTF(1, "Encountered error in abstract interpreter\n");
+    return ABSTRACT_INTERP_ERROR;
+
 }
 
+static int
+uop_abstract_interpret(
+    PyCodeObject *co,
+    _PyUOpInstruction *trace,
+    _PyUOpInstruction *new_trace,
+    int trace_len,
+    int curr_stacklen
+)
+{
+    bool did_loop_peel = false;
+
+    _Py_UOpsAbstractInterpContext *ctx = NULL;
+
+    ctx = abstractinterp_context_new(
+        co, curr_stacklen,
+        trace_len, new_trace);
+    if (ctx == NULL) {
+        goto error;
+    }
+    _PyUOpInstruction *curr = NULL;
+    _PyUOpInstruction *end = NULL;
+    AbstractInterpExitCodes status = ABSTRACT_INTERP_NORMAL;
+    bool first_impure = true;
+    bool has_enough_space_to_duplicate_loop = true;
+    int res = 0;
+
+loop_peeling:
+    curr = trace;
+    end = trace + trace_len;
+    first_impure = true;
+    ;
+    while (curr < end && !op_is_end(curr->opcode)) {
+
+        if (!op_is_pure(curr->opcode) &&
+            !op_is_specially_handled(curr->opcode) &&
+            !op_is_bookkeeping(curr->opcode) &&
+            !op_is_guard(curr->opcode)) {
+            DPRINTF(3, "Impure %s\n", (curr->opcode >= 300 ? _PyOpcode_uop_name : _PyOpcode_OpName)[curr->opcode]);
```

Comment:
I just realized that you can always use the uop name array here:
```suggestion
            DPRINTF(3, "Impure %s\n", _PyOpcode_uop_name[curr->opcode]);
```
Also at L684 and L790.